{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The time difference is : 0.05008139999972627\n"
     ]
    }
   ],
   "source": [
    "#PATH SETTINGS\n",
    "#--------------------------------------------------------------------------\n",
    "import os.path\n",
    "#NOTE: Slightly nicer way of handling paths\n",
    "#save_path = \"C:/Users/micha/OneDrive/GoodNotes/Projects/Protein_Identification\"\n",
    "#npep_dir = os.path.join(save_path, 'npep')\n",
    "#--------------------------------------------------------------------------\n",
    "\n",
    "\n",
    "#GENERATE PEPTIDES\n",
    "#--------------------------------------------------------------------------\n",
    "from itertools import combinations\n",
    "\n",
    "#list of aminoacids\n",
    "aacids = [\"A\",\"R\",\"N\",\"D\",\"C\",\"E\",\"Q\",\"G\",\"H\",\"I\",\"L\",\"K\",\"M\",\"F\",\"P\",\"S\",\"T\",\"W\",\"Y\",\"V\"]\n",
    "\n",
    "def generateCombination(chars, n):\n",
    "    \"\"\"\n",
    "    Generate all combinations of letters in chars of size n\n",
    "    \"\"\"\n",
    "    combilist = [''.join(comb) for comb in combinations(chars, n)]\n",
    "    return combilist\n",
    "\n",
    "#create npeps\n",
    "dipeps = generateCombination(aacids,2)\n",
    "tripeps = generateCombination(aacids,3)\n",
    "tetrapeps = generateCombination(aacids,4)\n",
    "\n",
    "#output location file header preparation\n",
    "#f_tripeps = open(os.path.join(npep_dir, \"tripeps.txt\"), \"w\")\n",
    "f_tripeps = open(\"C:\\\\Users\\\\sergi\\\\Desktop\\\\projects\\\\peptides\\\\tripep.txt\", \"w\") #ACHTUNG CHANGE PATH\n",
    "\n",
    "#fill file with combinations in proper format\n",
    "f_tripeps.write(\"name\")\n",
    "for j in tripep:\n",
    "    f_tripeps.write(\"\\t\")\n",
    "    f_tripeps.write(j)\n",
    "f_tripeps.write(\"\\t\")\n",
    "f_tripeps.write(\"total\")\n",
    "f_tripeps.write(\"\\n\")\n",
    "f_tripeps.close()\n",
    "\n",
    "#f_tripeps_tot = open(os.path.join(npep_dir, \"tripeps_tot.txt\"), \"w\")  #ACHTUNG CHANGE PATH\n",
    "f_tripeps_tot = open(\"C:\\\\Users\\\\sergi\\\\Desktop\\\\projects\\\\peptides\\\\tripep_tot.txt\", \"w\")  #ACHTUNG CHANGE PATH\n",
    "f_tripeps_tot.write(\"total\")\n",
    "for j in tripeps:\n",
    "    f_tripeps_tot.write(\"\\t\")\n",
    "    f_tripeps_tot.write(j)\n",
    "else: f_tripeps_tot.write(\"\\n\")\n",
    "f_tripeps_tot.close()\n",
    "#--------------------------------------------------------------------------\n",
    "\n",
    "\n",
    "#PREPARE PROTEOME\n",
    "#--------------------------------------------------------------------------\n",
    "# Open a file: file\n",
    "#f_proteome = open(os.path.join(save_path, \"uniproteomerev.fasta\"))\n",
    "f_proteome = open('C:\\\\Users\\\\sergi\\\\Downloads\\\\uniproteomerev.fasta' , \"r\")  #ACHTUNG CHANGE PATH\n",
    "proteome = f_proteome.read()\n",
    "f_proteome.close()\n",
    "#--------------------------------------------------------------------------\n",
    "\n",
    "\n",
    "#COUNT OCCURENCES\n",
    "#--------------------------------------------------------------------------\n",
    "import timeit\n",
    "import random\n",
    "\n",
    "#filling files\n",
    "#f_tripeps = open(os.path.join(npep_dir, \"tripeps.txt\"), \"a\")  #ACHTUNG CHANGE PATH\n",
    "f_tripeps = open(\"C:\\\\Users\\\\sergi\\\\Desktop\\\\projects\\\\peptides\\\\tripep.txt\", \"a\") #ACHTUNG CHANGE PATH\n",
    "\n",
    "#split and clip\n",
    "entries = proteome.split(\">\")\n",
    "entries = entries[1:]\n",
    "\n",
    "#time runtime of algorithm\n",
    "starttime = timeit.default_timer()\n",
    "\n",
    "#for each protein, count occurences of specific tripeptides\n",
    "for entry in entries:\n",
    "    #seperate entry format\n",
    "    prot = entry.split(\"\\n\", 1)\n",
    "    \n",
    "    #if input is not accessible, ignore\n",
    "    #NOTE: Here should be some sort of error handling - why does it not work sometimes?\n",
    "    try:\n",
    "        seq = prot[1].replace(\"\\n\", \"\")\n",
    "    except:\n",
    "        seq = \"\"\n",
    "    \n",
    "    #extract protein ID\n",
    "    prot_id = prot[0]\n",
    "    \n",
    "    #write ID to file\n",
    "    f_tripeps.write(prot_id)\n",
    "    \n",
    "    #counter for total number of occurences\n",
    "    tot = 0\n",
    "    \n",
    "    #count number of occurences of each motif in each protein\n",
    "    for tripep in tripeps:\n",
    "        #counts occurences of substring of another string\n",
    "        count = seq.count(tripep)\n",
    "        #count = len(re.findall(tripep,seq)) -> This caused it to be relatively slow\n",
    "        \n",
    "        #document results\n",
    "        f_tripeps.write(\"\\t\")\n",
    "        f_tripeps.write(str(count))\n",
    "        \n",
    "        #add to total occurences\n",
    "        tot += count\n",
    "        \n",
    "    #NOTE: Where is the if for this else?\n",
    "    else:\n",
    "        #\n",
    "        f_tripeps.write(\"\\t\")\n",
    "        f_tripeps.write (str(tot))\n",
    "        \n",
    "    #format\n",
    "    f_tripeps.write (\"\\n\")\n",
    "\n",
    "#close outstream\n",
    "f_tripeps.close()\n",
    "\n",
    "#print time it took\n",
    "print(\"The time difference is :\", timeit.default_timer() - starttime)\n",
    "#--------------------------------------------------------------------------"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
